source/utilities.tex

```diff
@@ -10244,11 +10358,26 @@
 
 \begin{itemdescr}
 \pnum
-\requires \tcode{Y} shall be a complete type. The expression
+\mandates
+\tcode{Y} is a complete type.
+
+\pnum
+\constraints
+When \tcode{T} is an array type,
+the expression \tcode{delete[] p} is well-formed and either
+\tcode{T} is \tcode{U[N]} and \tcode{Y(*)[N]} is convertible to \tcode{T*}, or
+\tcode{T} is \tcode{U[]} and \tcode{Y(*)[]} is convertible to \tcode{T*}.
+When \tcode{T} is not an array type,
+the expression \tcode{delete p} is well-formed and
+\tcode{Y*} is convertible to \tcode{T*}.
```

Comment:
I'm concerned this will change the presumptive intent. Consider:

```
#include <memory>
struct Base; struct Derived;
bool k = requires (Derived *p) { std::shared_ptr<Base*>(p); };
```

Does this violate a Mandates or a Constraints? (Put another way: is the above program guaranteed to be well-formed, or could it result in a non-immediate-context error?)

With the current order of elements, the suggestion (albeit a flawed one) is that the program can be rejected. Reversing the order of elements removes the suggestion that the *Mandates:* could somehow mysteriously apply prior to overload resolution, requiring that example to be accepted.